File: hs/Math/KnotTh/Tangles/TangleSkeleton.hs
module Math.KnotTh.Tangles.TangleSkeleton
	(
	  CTangleSkeleton
	, TangleSkeleton
	, DartSkeleton(DartSkeleton)
	, CrossingSkeleton(CrossingSkeleton)

	, wrapSkeleton
	, transformSkeleton
	, withSkeleton
	, skeletonNumberOfCrossings
	, skeletonNumberOfLegs
	, skeletonOpposite
	, skeletonNextCCW
	, skeletonNextCW
	) where

import qualified System.IO.Unsafe as IOUnsafe
import Foreign
import Foreign.C.Types


data CTangleSkeleton


foreign import ccall "_ZNK4Math6KnotTh7Tangles14TangleSkeleton18cNumberOfCrossingsEv"
	c_numberOfCrossings :: Ptr CTangleSkeleton -> IO CSize

foreign import ccall "_ZNK4Math6KnotTh7Tangles14TangleSkeleton13cNumberOfLegsEv"
	c_numberOfLegs :: Ptr CTangleSkeleton -> IO CSize

foreign import ccall "_ZNK4Math6KnotTh7Tangles14TangleSkeleton7cNextCWENS1_4DartE"
	c_nextCW :: Ptr CTangleSkeleton -> CSize -> IO CSize

foreign import ccall "_ZNK4Math6KnotTh7Tangles14TangleSkeleton8cNextCCWENS1_4DartE"
	c_nextCCW :: Ptr CTangleSkeleton -> CSize -> IO CSize

foreign import ccall "_ZNK4Math6KnotTh7Tangles14TangleSkeleton9cOppositeENS1_4DartE"
	c_opposite :: Ptr CTangleSkeleton -> CSize -> IO CSize

foreign import ccall "&_ZN4Math6KnotTh7Tangles14TangleSkeleton7releaseEv"
	c_release :: FunPtr (Ptr CTangleSkeleton -> IO ())


newtype TangleSkeleton = TangleSkeleton (ForeignPtr CTangleSkeleton)

newtype DartSkeleton = DartSkeleton CSize

newtype CrossingSkeleton = CrossingSkeleton CSize


instance Eq DartSkeleton where
	(==) (DartSkeleton a) (DartSkeleton b) = (a == b)

instance Eq CrossingSkeleton where
	(==) (CrossingSkeleton a) (CrossingSkeleton b) = (a == b)


instance Ord CrossingSkeleton where
	compare (CrossingSkeleton a) (CrossingSkeleton b) = compare a b


wrapSkeleton :: IO (Ptr CTangleSkeleton) -> TangleSkeleton
wrapSkeleton createAction =
	TangleSkeleton $! IOUnsafe.unsafePerformIO (createAction >>= newForeignPtr c_release)


transformSkeleton :: (Ptr CTangleSkeleton -> IO (Ptr CTangleSkeleton)) -> TangleSkeleton -> TangleSkeleton
transformSkeleton transformation (TangleSkeleton sk) =
	TangleSkeleton $! IOUnsafe.unsafePerformIO (withForeignPtr sk transformation >>= newForeignPtr c_release)


{-# INLINE withSkeleton #-}
withSkeleton :: (Ptr CTangleSkeleton -> IO a) -> TangleSkeleton -> a
withSkeleton action (TangleSkeleton sk) =
	IOUnsafe.unsafePerformIO (withForeignPtr sk action)


{-# INLINE skeletonNumberOfCrossings #-}
skeletonNumberOfCrossings :: TangleSkeleton -> Int
skeletonNumberOfCrossings (TangleSkeleton sk) =
	fromIntegral $! IOUnsafe.unsafePerformIO (withForeignPtr sk c_numberOfCrossings)


{-# INLINE skeletonNumberOfLegs #-}
skeletonNumberOfLegs :: TangleSkeleton -> Int
skeletonNumberOfLegs (TangleSkeleton sk) =
	fromIntegral $! IOUnsafe.unsafePerformIO (withForeignPtr sk c_numberOfLegs)


{-# INLINE skeletonOpposite #-}
skeletonOpposite :: TangleSkeleton -> DartSkeleton -> DartSkeleton
skeletonOpposite (TangleSkeleton sk) (DartSkeleton d) =
	DartSkeleton $! IOUnsafe.unsafePerformIO (withForeignPtr sk (\ ptr -> c_opposite ptr d))


{-# INLINE skeletonNextCCW #-}
skeletonNextCCW :: TangleSkeleton -> DartSkeleton -> DartSkeleton
skeletonNextCCW (TangleSkeleton sk) (DartSkeleton d) =
	DartSkeleton $! IOUnsafe.unsafePerformIO $ withForeignPtr sk (\ ptr -> c_nextCCW ptr d)


{-# INLINE skeletonNextCW #-}
skeletonNextCW :: TangleSkeleton -> DartSkeleton -> DartSkeleton
skeletonNextCW (TangleSkeleton sk) (DartSkeleton d) =
	DartSkeleton $! IOUnsafe.unsafePerformIO $ withForeignPtr sk (\ ptr -> c_nextCW ptr d)

File: hs/Math/KnotTh/Tangles/TangleSkeleton.cpp
#include "./TangleSkeleton.h"

namespace Math { namespace KnotTh { namespace Tangles {

	bool Dart::cIsLeg() const
	{
		return isLeg();
	}

	bool Dart::cIsDart() const
	{
		return isDart();
	}


	void TangleSkeleton::release()
	{
		delete this;
	}

	size_t TangleSkeleton::cNumberOfCrossings() const
	{
		return numberOfCrossings();
	}

	size_t TangleSkeleton::cNumberOfLegs() const
	{
		return numberOfLegs();
	}

	Dart TangleSkeleton::cDart(size_t crossing, size_t place) const
	{
		return dart(crossing, place);
	}

	Dart TangleSkeleton::cLeg(size_t l) const
	{
		return leg(l);
	}

	Dart TangleSkeleton::cOpposite(Dart dart) const
	{
		return opposite(dart);
	}

	Dart TangleSkeleton::cNextCCW(Dart dart) const
	{
		return nextCCW(dart);
	}

	Dart TangleSkeleton::cNextCW(Dart dart) const
	{
		return nextCW(dart);
	}

	int TangleSkeleton::cLexicographicalCompare(Dart a, Dart b) const
	{
		if(a.isLeg())
		{
			if(!b.isLeg())
				return -1;

			if(a.place < b.place)
				return -1;

			if(a.place > b.place)
				return 1;

			return 0;
		}
		else
		{
			if(b.isLeg())
				return 1;

			if(a.crossing < b.crossing)
				return -1;

			if(a.crossing > b.crossing)
				return 1;

			if(a.place < b.place)
				return -1;

			if(a.place > b.place)
				return 1;

			return 0;
		}
	}

	Dart TangleSkeleton::cLexicographicallyNext(Dart dart) const
	{
		if(dart.isLeg())
		{
			if(dart.place == numberOfLegs() - 1)
				return Dart(1, 0);
			else
				return Dart(0, dart.place + 1);
		}
		else
		{
			if(dart.place == 3)
				return Dart(dart.crossing + 1, 0);
			else
				return Dart(dart.crossing, dart.place + 1);
		}
	}

	size_t TangleSkeleton::cLexicographicalRangeSize(Dart a, Dart b) const
	{
		if(a.isLeg())
		{
			if(b.isLeg())
			{
				if(a.place <= b.place)
					return 1 + b.place - a.place;
				else
					return 0;
			}
			else
				return (numberOfLegs() - a.place) + (b.crossing * 4 + b.place - 3);
		}
		else
		{
			if(b.isLeg())
				return 0;
			else
			{
				const size_t x = a.crossing * 4 + a.place;
				const size_t y = b.crossing * 4 + b.place;
				return (x <= y ? 1 + y - x : 0);
			}
		}
	}

}}}
